library/math/num.lisp

```diff
@@ -161,6 +161,50 @@
 ;;; Num instances for integers
 ;;;
 
+(cl:defmacro %define-unsigned-handler (name bits)
+  `(cl:defun ,name (bits)
+     (cl:declare (type (unsigned-byte ,bits) bits))
+     (cl:ecase ,bits
+       (8 (cl:the (cl:unsigned-byte ,bits) 255))
+       (16 (cl:the (cl:unsigned-byte ,bits) 65535))
+       (32 (cl:the (cl:unsigned-byte ,bits) 4294967295))
+       (64 (cl:the (cl:unsigned-byte ,bits) 18446744073709551615)))))
+
+
+(%define-unsigned-handler %unsigned-8-bit-handler 8)
+(%define-unsigned-handler %unsigned-16-bit-handler 16)
+(%define-unsigned-handler %unsigned-32-bit-handler 32)
+(%define-unsigned-handler %unsigned-64-bit-handler 64)
+
+(cl:eval-when (:compile-toplevel :load-toplevel)
+  (cl:defmacro define-unsigned-num-underflow (type underflow-handler bits)
+    "Define a `Num' instance for Type which signals an error on underflow."
+    `(define-instance (Num ,type)
+       (define (+ a b)
+	 (lisp ,type (a b)
+	   (cl:cond ((cl:and (cl:minusp b) (cl:< a (cl:- b)))
+		     (cl:cerror "Continue, wrapping around." ,(cl:format cl:nil "Unsigned value underflowed ~D bits." bits))
+		     (,underflow-handler ,bits))
+		    (cl:+ a b))))
+	 
+	 (define (- a b)
+	   (lisp ,type (a b)
+	     (cl:cond ((cl:< a (cl:+ 0 b))
```

Comment:
what is the point of (cl:+ 0 b)? doesn't it just evaluate to b?

Reply:
I can simplify this. Thanks